File: SozlukJetpack/app/src/main/java/com/example/sozlukjetpack/view/EntryDetayScreen.kt
package com.example.sozlukjetpack.view

import android.widget.Toast
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Divider
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.viewmodel.compose.viewModel
import com.example.sozlukjetpack.model.Comment
import com.example.sozlukjetpack.util.SessionManager
import com.example.sozlukjetpack.viewmodel.EntryDetayViewModel


@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EntryDetayScreen(
    entryId: Int,
    session: SessionManager,
    vm: EntryDetayViewModel = viewModel(),
    onBack: () -> Unit
) {
    val context = LocalContext.current
    val entry by vm.entry.collectAsStateWithLifecycle()
    val comments by vm.comments.collectAsStateWithLifecycle()
    val ui by vm.ui.collectAsStateWithLifecycle()

    var commentText by remember { mutableStateOf("") }
    var voteDialogFor by remember { mutableStateOf<Comment?>(null) }

    LaunchedEffect(entryId) {
        vm.loadEntry(entryId)
        vm.loadComments(entryId)
    }

    LaunchedEffect(ui.error) {
        ui.error?.let { Toast.makeText(context, it, Toast.LENGTH_SHORT).show() }
    }

    Scaffold(
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text(entry?.title ?: "Entry Detay") },
                navigationIcon = {
                    TextButton(onClick = onBack) { Text("Geri") }
                }
            )
        }
    ) { padding ->
        Column(
            Modifier.padding(padding).padding(16.dp).fillMaxSize()
        ) {
            // Entry içerik
            if (ui.loadingEntry) {
                Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator()
                }
            } else entry?.let { e ->
                Text(text = e.title, style = MaterialTheme.typography.titleLarge)
                Spacer(Modifier.height(6.dp))
                Text(text = e.content, style = MaterialTheme.typography.bodyLarge)
                Spacer(Modifier.height(4.dp))
                val meta = listOfNotNull(e.username, e.created_at.takeIf { it.isNotBlank() }?.take(10)).joinToString(" • ")
                if (meta.isNotBlank()) Text(meta, style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            }

            Spacer(Modifier.height(16.dp))
            Divider()
            Spacer(Modifier.height(8.dp))
            Text("Yorumlar", style = MaterialTheme.typography.titleMedium)
            Spacer(Modifier.height(8.dp))

            if (ui.loadingComments) {
                Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator()
                }
            } else {
                if (comments.isEmpty()) {
                    Text("Henüz yorum yok")
                } else {
                    LazyColumn(verticalArrangement = Arrangement.spacedBy(8.dp), modifier = Modifier.weight(1f, fill = false)) {
                        items(comments, key = { it.id }) { c ->
                            ElevatedCard(Modifier.fillMaxWidth().clickable { voteDialogFor = c }) {
                                Column(Modifier.padding(12.dp)) {
                                    Text(c.username, style = MaterialTheme.typography.labelLarge)
                                    Spacer(Modifier.height(4.dp))
                                    Text(c.comment_text, style = MaterialTheme.typography.bodyMedium)
                                    Spacer(Modifier.height(6.dp))
                                    Text("👍${c.likes}  👎${c.dislikes}", style = MaterialTheme.typography.labelMedium)
                                }
                            }
                        }
                    }
                }
            }

            Spacer(Modifier.height(12.dp))
            OutlinedTextField(
                value = commentText,
                onValueChange = { commentText = it },
                label = { Text("Yorum yaz") },
                keyboardOptions = KeyboardOptions.Default,
                modifier = Modifier.fillMaxWidth()
            )
            Spacer(Modifier.height(8.dp))
            Button(
                onClick = {
                    if (commentText.isBlank()) return@Button
                    val userId = session.getUserId()
                    vm.addComment(entryId, userId, commentText)
                    commentText = ""
                },
                enabled = !ui.posting,
                modifier = Modifier.fillMaxWidth()
            ) {
                if (ui.posting) {
                    CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(20.dp))
                    Spacer(Modifier.width(8.dp))
                    Text("Gönderiliyor…")
                } else Text("Gönder")
            }
        }
    }

    voteDialogFor?.let { c ->
        AlertDialog(
            onDismissRequest = { voteDialogFor = null },
            title = { Text("Yorumu Oyla") },
            text = { Text(c.comment_text) },
            confirmButton = {
                TextButton(onClick = {
                    vm.voteComment(entryId, c.id, session.getUserId(), true)
                    voteDialogFor = null
                }) { Text("👍 Beğen") }
            },
            dismissButton = {
                TextButton(onClick = {
                    vm.voteComment(entryId, c.id, session.getUserId(), false)
                    voteDialogFor = null
                }) { Text("👎 Beğenme") }
            }
        )
    }
}

File: sozluk_flutter/lib/view/entrydetay_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../entity/session_manager.dart';
import '../model/comment.dart';
import '../viewmodel/entrydetay_viewmodel.dart';


class EntryDetayScreen extends StatefulWidget {
  final int entryId;
  final SessionManager session;
  final VoidCallback onBack;

  const EntryDetayScreen({
    super.key,
    required this.entryId,
    required this.session,
    required this.onBack,
  });

  @override
  State<EntryDetayScreen> createState() => _EntryDetayScreenState();
}

class _EntryDetayScreenState extends State<EntryDetayScreen> {
  bool _loaded = false;
  final _commentCtrl = TextEditingController();

  @override
  void dispose() {
    _commentCtrl.dispose();
    super.dispose();
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (!_loaded) {
      _loaded = true;
      final vm = context.read<EntryDetayViewModel>();
      vm.loadEntry(widget.entryId);
      vm.loadComments(widget.entryId);
    }
  }

  Future<int> _userId() => widget.session.getUserId();

  @override
  Widget build(BuildContext context) {
    final vm = context.watch<EntryDetayViewModel>();
    final ui = vm.ui;

    WidgetsBinding.instance.addPostFrameCallback((_) {
      final err = ui.error;
      if (err != null && err.trim().isNotEmpty) {
        ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text(err)));
      }
    });

    final entry = vm.entry;

    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text(entry?.title ?? 'Entry Detay'),
        leading: TextButton(onPressed: widget.onBack, child: const Text('Geri')),
        leadingWidth: 72,
      ),
      body: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          children: [
            if (ui.loadingEntry)
              const Center(child: CircularProgressIndicator())
            else if (entry != null)
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(entry.title, style: Theme.of(context).textTheme.titleLarge),
                  const SizedBox(height: 6),
                  Text(entry.content, style: Theme.of(context).textTheme.bodyLarge),
                  const SizedBox(height: 6),
                  Text(
                    [
                      entry.username,
                      entry.createdAt.isNotEmpty ? entry.createdAt.substring(0, entry.createdAt.length >= 10 ? 10 : entry.createdAt.length) : null
                    ].whereType<String>().join(' • '),
                    style: Theme.of(context).textTheme.bodySmall?.copyWith(
                      color: Theme.of(context).colorScheme.onSurfaceVariant,
                    ),
                  ),
                ],
              ),

            const SizedBox(height: 16),
            const Divider(),
            const SizedBox(height: 8),
            Align(
              alignment: Alignment.centerLeft,
              child: Text('Yorumlar', style: Theme.of(context).textTheme.titleMedium),
            ),
            const SizedBox(height: 8),
            
            Expanded(
              child: Builder(
                builder: (_) {
                  if (ui.loadingComments) {
                    return const Center(child: CircularProgressIndicator());
                  }
                  final comments = vm.comments;
                  if (comments.isEmpty) {
                    return const Center(child: Text('Henüz yorum yok'));
                  }
                  return ListView.separated(
                    itemCount: comments.length,
                    separatorBuilder: (_, __) => const SizedBox(height: 8),
                    itemBuilder: (_, i) {
                      final c = comments[i];
                      return Card(
                        elevation: 1,
                        child: InkWell(
                          onTap: () => _openVoteDialog(context, c),
                          borderRadius: BorderRadius.circular(12),
                          child: Padding(
                            padding: const EdgeInsets.all(12),
                            child: Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(c.username, style: Theme.of(context).textTheme.labelLarge),
                                const SizedBox(height: 4),
                                Text(c.commentText),
                                const SizedBox(height: 8),
                                Text('👍${c.likes}   👎${c.dislikes}',
                                    style: Theme.of(context).textTheme.labelMedium),
                              ],
                            ),
                          ),
                        ),
                      );
                    },
                  );
                },
              ),
            ),

            const SizedBox(height: 12),
            TextField(
              controller: _commentCtrl,
              decoration: const InputDecoration(
                labelText: 'Yorum yaz',
                border: OutlineInputBorder(),
              ),
            ),
            const SizedBox(height: 8),

            SizedBox(
              width: double.infinity,
              child: ElevatedButton(
                onPressed: ui.posting
                    ? null
                    : () async {
                  final text = _commentCtrl.text.trim();
                  if (text.isEmpty) return;

                  final userId = await _userId();
                  await context.read<EntryDetayViewModel>().addComment(
                    entryId: widget.entryId,
                    userId: userId,
                    text: text,
                  );
                  _commentCtrl.clear();
                },
                child: ui.posting
                    ? const SizedBox(
                  height: 20, width: 20, child: CircularProgressIndicator(strokeWidth: 2),
                )
                    : const Text('Gönder'),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Future<void> _openVoteDialog(BuildContext context, Comment c) async {
    final userId = await _userId();

    if (!context.mounted) return;

    showDialog(
      context: context,
      builder: (_) => AlertDialog(
        title: const Text('Yorumu Oyla'),
        content: Text(c.commentText),
        actions: [
          TextButton(
            onPressed: () async {
              Navigator.pop(context);
              await context.read<EntryDetayViewModel>().voteComment(
                entryId: widget.entryId,
                commentId: c.id,
                userId: userId,
                isLike: false,
              );
            },
            child: const Text('👎 Beğenme'),
          ),
          TextButton(
            onPressed: () async {
              Navigator.pop(context);
              await context.read<EntryDetayViewModel>().voteComment(
                entryId: widget.entryId,
                commentId: c.id,
                userId: userId,
                isLike: true,
              );
            },
            child: const Text('👍 Beğen'),
          ),
        ],
      ),
    );
  }
}
